### ISSUE TITLE
[EPIC] Create an eCR data store for LAC

### ISSUE THREAD

**Issue by emmastephenson** [CDCgov/phdi#191]
# Why are we building this?
One of our goals for our pilot with LAC is to make eCR usable in IRIS.

To be able to have the pipeline we've set up for LAC integrate in any useful way with IRIS, we need to give LAC a tsv extract because it's one of the ways they can get data into IRIS. See these notes [here](https://docs.google.com/document/d/13pO-U6X2ruD-cOg3ku9u-Cc9afV9Y-GyGknClY7uGZ4/edit#) for information about LAC's import utility and how we decided on using a tsv to get data into IRIS. 

We are focusing on integrating eCR data into IRIS because LAC cannot currently get any eCR data into IRIS, meaning eCR is not usable for them at all right now.

To be able to give LAC the tsv extract they need to integrate our pipeline with IRIS, and therefore be able to process eCR, we need to create a data store for them. We cannot provide a tsv extract to them until we set up the data store on our end. This epic covers the work we need to do on our end before we can even think about using a tsv extract to integrate with IRIS. 

#### _What problem does this solve for the user?_

This epic is the first step towards integrating eCR data into IRIS, which would then enable LAC to use eCR in a meaningful way. But to fully solve this problem, we will need the following epic around extracting the data from the data store to integrate with IRIS.

This work also is able to solve two secondary problems for LAC: 

- LAC indicated it would be useful to have a data store outside of their disease surveillance system, IRIS, so that they can perform analytics on their data without needing to go through IRIS

- By setting up a message parsing service, we will solve provenance tracking for eCR in the short run. With our existing pipeline, LAC cannot distinguish which data came from eCR/ELR/etc. While we intend to implement a provenance resource to solve this problem in the long run, creating a table to store LAC's eCR data will solve this problem in the short run before the pilot is over. Note: this problem might not even be a problem to solve for LAC in our pilot because we may not be able to run ELR data through the pipeline anyways, so all data would be eCR data in that case.

#### _Relevant context or background info_
- [Notes from LAC Systems Sync 2/1-2/17](https://docs.google.com/document/d/13pO-U6X2ruD-cOg3ku9u-Cc9afV9Y-GyGknClY7uGZ4/edit#)
- [Notes from LAC Systems Sync 3/1-3/15](https://docs.google.com/document/d/1i6wfU1llCQ9-X96qpKWC9rHziRSo_nNXgD-YTHoBZzo/edit#heading=h.oooyyeuw1bk3)
- What is provenance tracking?
  - More background info on what provenance tracking is and why it's important is available from google cloud [here](https://cloud.google.com/architecture/tracking-provenance-and-lineage-metadata-for-healthcare-data). You can also look at the documentation for the [FHIR provenance resource](https://fhir-ru.github.io/provenance.html) directly.

# Acceptance Criteria
**GIVEN** an eCR  message enters our ingestion pipeline in LAC Azure
**WHEN** the message is processed through the pipeline
**THEN** the data fields that are important to LAC are extracted from the eCR message and are written to a persistent tabular data store

# Engineering Solution
To be able to implement a data store for LAC for eCR, we broadly need to (1) create a message parsing service that extracts LAC's required fields for eCR data from our pipeline, (2) integrate our message parsing service into our pipeline, and (3) set up a data store to hold eCR data and write data into it.

#### Message parsing service

We need to set up a message parsing service because we need to extract specific fields from any given message. Being able to extract specific fields from messages will allow us to get the fields that LAC requires for an tsv extract out of eCR data coming into our pipeline. 

Specific tickets:
- Implement a message parsing service
  - How it works: Give it a mapping between the fields you want and our FHIR bundle, this message service will determine the paths needed and extract resources from the FHIR bundle. This is what we'll use to extract data out of eCR instead of using the tabulation service, which is not ready.

#### Integrating message parsing service into our pipeline

Once we implement a message parsing service, we then need to integrate it into our pipeline generally, which means we have to containerize it and make it available as a Building Block in our starter kit. We also specifically need to set up the message parsing service to LAC specifications to integrate it into the pipeline we have set up for them.

Specific tickets:

- Add message parsing service to starter kit
  - The message parsing service should deploy as part of the starter kit along with the other DIBBs containerized services
- Create message parsing schema for eCR messages according to LAC specification
  - Review the fields that LAC requires from eCR, determine the FHIRpath for each of these fields in the FHIR bundles resulting from the conversion of an eCR from XML to FHIR, and update the ecr.json schema in the message parsing service with these field:FHIRpath key:value pairs.
- Modify ingestion pipeline to parse eCR messages with the message parsing service and publish to event hub
  - When an eCR message moves through the pipeline we need to use the message parsing service to extract the fields that LAC wants from the message. Once these fields have been extracted the data should be published to an Azure Event Hub topic.
- Add step to pipeline to publish parsed messaged to Azure Event Hub topic

#### Setting up data store and writing data to it

We need to set up a data store to hold the eCR data from the message parsing service set up to LAC's specs, and then we need to actually write eCR data to the table

Specific tickets:
- Determine table schema for eCR data store
  - This schema will likely be informed by the sample tsv that LAC passes to our team, but should be flexible enough to incorporate additional data fields in the future if they are requested
- Create storage resource to hold eCR records
  - As an IT engineer, I need to pass tsv eCR data that matches the IRIS required fields to LAC’s import utilities to seed IRIS and make that data available to case managers. The easiest way for the DIBBs pipeline to support this is through the creation of an eCR database.
  - Determine the database type best suited for this eCR database (Postgres? Another DB?)
  - Create the resources needed to support this type of resource in Azure.
  - See discussion in [this document](https://docs.google.com/document/d/1-prnFmiSEStdZjJT-XjAFXda_3p1AWwSaacpB_PtZZg/edit?usp=sharing) for additional details.
- Create a Spark streaming job to read parsed eCR data from Azure Event Hub and write to delta table
  - After parsed eCR data is published to Azure Event Hub we need a way to read this data and write it as rows to an eCR Delta Table. AC: When the stream job runs all available data on the Azure Event Hub topic is read and appended to the eCR delta table.
- Update starter kit phdi storage account to ADLS Gen 2
  - In order to support Delta Tables in our storage accounts we need to upgrade to ADLS Gen2. this can be by enable "hierarchical name spaces" on our storage accounts. AC: The PHI storage account in the starter kit has been upgraded to ADLS Gen2

Last step:
- Publish changes to LAC's pipeline

# Future Work
This epic will be followed by an epic to extract data from the data store to seed eCR data into IRIS.

**Comment by aathwal3** [comment]
**Sprint Backlog**
- eCR messages can be extracted and stored in Azure data hub

**Comment by emmastephenson** [comment]
Goals for next sprint: We're generating the eCR data store.

**Comment by emmastephenson** [comment]
Update 5/1: Delta tables are being created with stubbed data.

**Comment by emmastephenson** [comment]
Nick Bristow and Dan will work on getting the eCR data store actually written in Azure.

Next sprint: it's all writing end to end in Azure.

**Comment by emmastephenson** [comment]
Dan and Nick B continuing to work on the Azure issues with the eCR data store.

Reading from Azure event hub ok but having issues with storage.

**Comment by emmastephenson** [comment]
This is complete!
